Java frontend: clean up unused clinit symbols

smowton opened the pull request:
Previously clinit wrappers that had been generated but never actually called would
remain in the GOTO program as empty, uncalled functions. Now they are cleaned up
in the same way as unused user functions.

smowton committed:
Java frontend: clean up unused clinit symbols

Previously clinit wrappers that had been generated but never actually called would
remain in the GOTO program as empty, uncalled functions. Now they are cleaned up
in the same way as unused user functions.

thk123 commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -196,7 +198,8 @@ bool ci_lazy_methodst::operator()(
       if(
-        method_bytecode.contains_method(sym.first) &&
+        (method_bytecode.contains_method(sym.first) ||
+         synthetic_methods.count(sym.first)) &&
```
When is this first condition false for symbol entries that are code (surely it is either a user function or a synthetic function?)

smowton replied:
This could be read as `is-normal-method || is-synthetic-method`, which I think matches your expectation.

thk123 replied:
Yes - but when is the above condition false?

smowton replied:
Ah, when the method belongs to a different frontend. For Java that case is currently unsupported, but cross-language or GOTO-binary + Java or similar possibilities are supported by our current architecture, so it would be best not to rule that case out when it's easy to support.

Better summary of this line: "if this method relates to this languaget" (otherwise leave it alone)

thk123 replied:
Ah yes - which matches with the comment :+1:

thk123 reviewed (commented):
A (unit) test showing clinit methods being correctly removed for:

 - simple unused clinit method
 - chained clinit method
 - stub clinit methods (that's the other kind of synthetic method right?)

smowton committed:
Add tests to verify that synthetic method removal is performed

These check that direct and indirect clinit wrappers, as well as stub class synthetic
static initialisers, are removed if they are generated but not subsequently called.

smowton commented:
@thk123 ready for re-review, tests added as requested.
